onnxruntime/featurizers_ops/cpu/short_grain_dropper_transformer.cc

```diff
@@ -0,0 +1,68 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/common.h"
+#include "core/framework/op_kernel.h"
+
+#include "Featurizers/ShortGrainDropperFeaturizer.h"
+#include "Featurizers/../Archive.h"
+
+namespace NS = Microsoft::Featurizer;
+
+namespace onnxruntime {
+namespace featurizers {
+
+void ShortGrainDropperTransformerImpl(OpKernelContext* ctx) {
+  // Create the transformer
+  Microsoft::Featurizer::Featurizers::ShortGrainDropperTransformer transformer(
+      [ctx]() {
+        const auto* state_tensor(ctx->Input<Tensor>(0));
+        const uint8_t* const state_data(state_tensor->Data<uint8_t>());
+
+        Microsoft::Featurizer::Archive archive(state_data, state_tensor->Shape().GetDims()[0]);
+        return Microsoft::Featurizer::Featurizers::ShortGrainDropperTransformer(archive);
+      }());
+
+  // Get the input
+  const auto* input_tensor = ctx->Input<Tensor>(1);
+  const std::string* input_data = input_tensor->template Data<std::string>();
+
+  // Prepare the output
+  const int64_t input_rows_num = input_tensor->Shape()[0];
+  const int64_t strings_num = input_tensor->Shape()[1];
+  TensorShape rows_shape({input_rows_num});
+  Tensor* output_tensor(ctx->Output(0, rows_shape));
+  bool* output_data(output_tensor->MutableData<bool>());
+
+  // Transform
+  for (int64_t rows_idx = 0; rows_idx < input_rows_num; ++rows_idx) {
+    std::vector<std::string> input_data_vec;
+    input_data_vec.reserve(strings_num);
+    for (int64_t string_idx = 0; string_idx < strings_num; ++string_idx)
+      input_data_vec.emplace_back(std::move(input_data[string_idx + rows_idx * strings_num]));
```

Comment:
>string_idx + rows_idx * strings_num [](start = 55, length = 35)

I suggest use and advance a single pointer to the input. Avoid indexing, it is slow.

Reply:
Here is a suggestion. Have a single ptr and remove the inner for. Replace it with the following (assuming input_data_vec is out of loops):
input_data_vec.clear();
input_data_ptr = std::copy(input_data_ptr, input_data_ptr + strings_num, std::back_inserter(input_data_vec));


---
In reply to: [393943438](https://github.com/microsoft/onnxruntime/pull/3241#discussion_r393943438) [](ancestors = 393943438)